service/history/handler.go

```diff
@@ -1238,6 +1272,49 @@ func (h *Handler) SyncActivity(ctx context.Context, syncActivityRequest *hist.Sy
 	return nil
 }
 
+// GetReplicationTasks is called by remote peers to get replicated tasks for cross DC replication
+func (h *Handler) GetReplicationTasks(
+	ctx context.Context,
+	request *r.GetReplicationTasksRequest,
+) (*r.GetReplicationTasksResponse, error) {
+
+	var wg sync.WaitGroup
+	wg.Add(len(request.Tokens))
+	result := new(sync.Map)
+
+	for _, token := range request.Tokens {
+		go func(token *r.ReplicationToken) {
+			defer wg.Done()
+
+			engine, err := h.controller.getEngineForShard(int(*token.ShardID))
+			if err != nil {
+				h.GetLogger().Warn("history engine not found for shard", tag.Error(err))
+				return
+			}
+
+			tasks, err := engine.GetReplicationTasks(ctx, *token.TaskID)
+			if err != nil {
+				h.GetLogger().Warn("failed to get replication tasks for shard", tag.Error(err))
```

Comment:
Is this a Warn or Error log? In which case this is returnning error but is expected? like the previous one history engine not found because of shard movement.

Reply:
It's not expected, but we ignore them and the expectation is that we will try again for the next request.

Reply:
We can retry but getting errors from here usually means that it won't recover. If that's true let's use error.

Reply:
What do you mean by it won't recover? I only expect transient errors in this case and they should recover in the next retry.